### ISSUE TITLE
Warning action executor is not executing.

### ISSUE THREAD

**Issue by mibby** [ruany/LiteBans#38]
@ruany Since updating to LiteBans 2.2.5 from version 2.2.4, I've experienced a couple of issues.

When a player reaches the defined amount of needed warnings for an action to occur, the action does not actually execute.

![link](https://i.imgur.com/okKuq20.png)

```
warnings:
  # Warnings will expire after this duration has passed.
  expire_after: 7 days

  # Warning actions.
  # Format: <amount>[+]:<action>
  # If '+' is provided after the amount, the action will execute if the player has that amount of warnings or more.
  actions:
  - '3:/tempban -s:$silent $player 1 day Reached 3 warnings: $list'

  # If enabled, the console will execute warning actions.
  # Otherwise, the player who executes the last warning also executes the warning action.
  actions_execute_as_console: false

  # If enabled, all active warnings will be deactivated after a warning action is executed.
  delete_warnings_after_action: false
```

The player received the warning and it says they were temp-banned for 1 day, but they aren't kicked or temp-banned. It effectively doesn't do anything.

```
# If bungeecord is set to true, banned players will be kicked from BungeeCord itself.
# (Players will not be kicked to the lobby when they are banned.)
# This option has no effect if the plugin is running directly under BungeeCord.
bungeecord: true
```

---

The last warning received says (forever ago) instead of an actual date (moments ago). I'm assuming it blanket defaults to forever ago until you look up how long it actually was? It shouldn't say forever ago to the player in my opinion when it was just executed.

**Comment by ruany** [comment]
Messages.yml does not include the time since warnings were placed anymore in the $list variable.
`    list_entry: ' \- Warned by $executor: ''$reason&f'''`

I can't replicate the executor issue, it doesn't happen for me.

**Comment by ruany** [comment]
Closing since actions work on my end.

**Comment by mibby** [comment]
@ruany It's reproducible 100% of the time for me. Perhaps you are not testing with BungeeCord setting enabled for it to kick players through plugin channels to Bungee?

LiteBans is on all bukkit servers with the bungeecord setting enabled. It is not on bungeecord.

**Comment by ruany** [comment]
Just tested on a local Spigot+Bungee setup with version 2.2.5 and `bungeecord: true`.

Spigot console:
```
>litebans:kick ruan
[05:00:35 INFO]: Ruan was kicked by Console.
[05:00:35 INFO]: [LiteBans] Kicking Ruan from BungeeCord.
[05:00:35 INFO]: Ruan lost connection: Disconnected
[05:00:35 INFO]: Ruan left the game
```

Bungee console:
```
05:00:22 [INFO] [Ruan] <-> ServerConnector [lobby] has connected
05:00:35 [INFO] [Ruan] disconnected with: Kicked by Console.
05:00:35 [INFO] [Ruan] -> UpstreamBridge has disconnected
05:00:35 [INFO] [Ruan] <-> DownstreamBridge <-> [lobby] has disconnected
```

Works as expected.

**Comment by mibby** [comment]
@ruany I don't use the kick warning command.

This specifically I removed.
```
  - '3:/kick -s:$silent $player Final warning: $reason'
```

This specifically is all I have.
```
  - '3:/tempban -s:$silent $player 1 day Reached 3 warnings: $list'
```

Perhaps what broke is the direct tempban command executed when receiving a number of warnings isn't kicking players anymore?

**Comment by ruany** [comment]
There is no difference between executing commands as a player and as a warning action, there is no special handling in place for commands handled by LiteBans or otherwise.

Regardless:
```
...
[01:17:54 INFO]: Console warned Ruan for 'test4'
[01:17:54 INFO]: Console tempbanned Ruan for 1 day for 'Reached 4 warnings: 
 - Warned by Console: 'test1'
 - Warned by Console: 'test2'
 - Warned by Console: 'test3'
 - Warned by Console: 'test4''
[01:17:54 INFO]: [LiteBans] Kicking Ruan from BungeeCord.
[01:17:54 INFO]: Ruan lost connection: Disconnected
[01:17:54 INFO]: Ruan left the game
```

**Comment by mibby** [comment]
@ruany I'm really not sure what to say. Since updating, no matter what I do, the warning system does not execute the actions for me. With ```actions_execute_as_console``` set to true and set to false, with ```bungeecord``` set to true and false, and even testing by warning from console instead of in-game. The player gets the warnings but they aren't temp-banned.

I even re-added the kick execution before the temp-ban in case that was the issue, but it didn't kick the player either.

2.2.4 works perfectly, 2.2.5 doesn't. I did see that you ninja uploaded 2.2.6 inside the 2.2.5 update so I'll give that a try as well.

Is there any way to debug the entire process chain to figure out why it isn't executing for me from LiteBans?

**Comment by mibby** [comment]
@ruany Apologies for the late replies, work has been keeping me preoccupied! 

Yes, kicking and banning players manually from in-game and from console both work fine for me. It's just the warning actions that aren't executing / working.

**Comment by ruany** [comment]
Do warning actions work on a fresh testing server with no other plugins installed?

**Comment by mibby** [comment]
@ruany After rigorous testing and cloning the server locally, I believe I finally figured out what may be the problem. I believe LiteBans is getting and using the displayname and/or color codes / prefixes of a player to kick them rather than using their raw playername from in-game / BungeeCord.

```
[17:15:08] [pool-8-thread-6/INFO]: An internal error occurred while attempting to perform this command.
[17:15:08] [pool-8-thread-6/WARN]: java.lang.IllegalArgumentException: Player name "SecurityBunny" length > 16
[17:15:08] [pool-8-thread-6/WARN]: 	at litebans.bR.b(bR.java:167)
[17:15:08] [pool-8-thread-6/WARN]: 	at litebans.bR.b(bR.java:348)
[17:15:08] [pool-8-thread-6/WARN]: 	at litebans.cO.a(cO.java:188)
[17:15:08] [pool-8-thread-6/WARN]: 	at litebans.cT.b(cT.java:52)
[17:15:08] [pool-8-thread-6/WARN]: 	at litebans.cp.run(cp.java:2)
[17:15:08] [pool-8-thread-6/WARN]: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
[17:15:08] [pool-8-thread-6/WARN]: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
[17:15:08] [pool-8-thread-6/WARN]: 	at java.lang.Thread.run(Unknown Source)
```

The playername was ```SecurityBunny``` with a blue nametag in-game with the color code &b. Their name is 13 characters normally, so shouldn't be >16 with &b prefix. Because of this error, this person does not get kicked or banned by LiteBans when relogging after the 3 warnings.

I typed;
 ```/warn SecurityBunny test1```
_Waited 30 seconds_
 ```/warn SecurityBunny test2```
_Waited 30 seconds_
 ```/warn SecurityBunny test3```
Error in console appeared.

Testing with my other account which has a 12 character name (14 with the prefix &b), I was able to get all 3 warnings to go off without the console exception occurring. However upon initiating the 3rd warning and what should have kicked the player from the server, it didn't. Outputs they have been temporarily banned but they aren't kicked. Relogging with the account warned 3 times, they are indeed temp-banned. Thus I believe LiteBans is trying to kick someone with a color code / prefix name, which doesn't exist, rather than the actual player.

```
[17:25:30] [main/INFO]: [CHAT] -----------------------------------------
[17:25:30] [main/INFO]: [CHAT]  » <player>has been warned by SecurityBunny
[17:25:30] [main/INFO]: [CHAT]  » Reason: test1
[17:25:30] [main/INFO]: [CHAT] -----------------------------------------
[17:26:41] [main/INFO]: [CHAT] -----------------------------------------
[17:26:41] [main/INFO]: [CHAT]  » <player>has been warned by SecurityBunny
[17:26:41] [main/INFO]: [CHAT]  » Reason: test2
[17:26:41] [main/INFO]: [CHAT] -----------------------------------------
[17:27:13] [main/INFO]: [CHAT] -----------------------------------------
[17:27:13] [main/INFO]: [CHAT]  » <player> has been warned by SecurityBunny
[17:27:13] [main/INFO]: [CHAT]  » Reason: test3
[17:27:13] [main/INFO]: [CHAT] -----------------------------------------
[17:27:14] [main/INFO]: [CHAT] -----------------------------------------
[17:27:14] [main/INFO]: [CHAT]  » <player> has been temporarily banned by Console
[17:27:14] [main/INFO]: [CHAT]  » Reason: Reached 3 warnings: 
[17:27:14] [main/INFO]: [CHAT]  \- Warned by SecurityBunny: 'test1'
[17:27:14] [main/INFO]: [CHAT]  \- Warned by SecurityBunny: 'test2'
[17:27:14] [main/INFO]: [CHAT]  \- Warned by SecurityBunny: 'test3'
[17:27:14] [main/INFO]: [CHAT]  » Duration: 1 day
[17:27:14] [main/INFO]: [CHAT] -----------------------------------------
```

Not kicked.

Removing the prefix from the player and player's group, the plugin ```Vault```, the plugin ```ColoredTags``` (which I use to colorize the displayname tag above player heads), the plugin ```ChatControlPro``` (which I use to handle chat formatting in-game), and even the permissions plugin itself, the player is still _not_ able to be kicked.

It wasn't until I finally removed EssentialsX that I was able to be kicked from the server / BungeeCord again after receiving 3 warnings.

Using EssentialsX dev 575.
https://ci.destroystokyo.com/view/all/job/EssentialsX/
https://github.com/EssentialsX/Essentials

I use the main EssentialsX.jar, EssentialsXAntiBuild.jar, and EssentialsSpawnX.jar files.

I am unsure if these settings from EssentialsX are what is affecting it or if it is a plugin conflict in general, but here are the relevant displayname settings I use from Essentials. I stopped testing once I located the conflict was between Essentials and LiteBans.

```
# A color code between 0-9 or a-f. Set to 'none' to disable.
ops-name-color: 'none'

# The character(s) to prefix all nicknames, so that you know they are not true usernames.
nickname-prefix: '~'

# The maximum length allowed in nicknames. The nickname prefix is included in this.
max-nick-length: 15

# When this option is enabled, nickname length checking will exclude color codes in player names.
# ie: "&6Notch" has 7 characters (2 are part of a color code), a length of 5 is used when this option is set to true
ignore-colors-in-max-nick-length: true

# Disable this if you have any other plugin, that modifies the displayname of a user.
change-displayname: true

# When this option is enabled, the (tab) player list will be updated with the displayname.
# The value of change-displayname (above) has to be true.
change-playerlist: true

# When EssentialsChat.jar isn't used, force essentials to add the prefix and suffix from permission plugins to displayname.
# This setting is ignored if EssentialsChat.jar is used, and defaults to 'true'.
# The value of change-displayname (above) has to be true.
# Do not edit this setting unless you know what you are doing!
add-prefix-suffix: true

# When this option is enabled, player prefixes will be shown in the playerlist.
# This feature only works for Minecraft version 1.8 and higher.
# This value of change-playerlist has to be true
add-prefix-in-playerlist: true
```

I'm fairly certain this use to work with LiteBans 2.2.4 and only broke upon updating to 2.2.5/2.2.6, as I recall it functioning fine when initially testing after purchasing the plugin. Hopefully this information will help you figure out what changed so it can be rectified. 

---

Second issue I noticed is once a player has received 3 warnings, they will never be kicked again with the action specified until all their current warnings expire. Such as;

* A player received 3 warnings,
* Gets temp-banned for 24 hours
* Comes back after a day, gets warned again 3 times. However isn't kicked because he now has 6 warnings due to the warning expiration lasting 7 days.

Thus is it possible to have it kick a player _every_ 3 warnings they receive and be able to list the last 3 warnings received when kicking / banning rather than all the warnings? This way you could keep an X day history.

---

Third issue, very minor. When I was diagnosing the kick issue locally, I switched LiteBans over to an H2 database. When I accumulated 3 warnings for the first time and it didn't kick me, I unwarned via ```/unwarn <name>``` until I had no more warnings. Then when I warned myself 3 more times, it always outputted ```Previous ban for <player> removed.``` on the 3rd warning. Not sure if it is an issue with H2 database where it resets any previous bans to re-ban, but I just started deleting the database file manually on every re-test.

**Comment by ruany** [comment]
1)
If it were retrieving the display name, it would say "§bSecurityBunny" not "SecurityBunny". I can only guess that invisible characters are somehow involved if the string length > 16.

I'll have to do my own testing with display names and the latest EssentialsX, but LiteBans shouldn't ever use display names in place of player names, otherwise you would see display names in the database (history table)

2) Seems a bit too complex, you should duplicate the third warning action as the sixth and give it a longer ban duration e.g. 48 hours (same with ninth, twelfth if warnings are really that common)
Creating new problems for me to solve here.

3) Expected behavior, new bans override existing ones

**Comment by ruany** [comment]
I still can't replicate the bug using latest EssentialsX+EssentialsXChat.

Tried usernames "/nick &a&b&c&d&1&2&3&4&5&6&7&8&fRuan" and "/nick ImpossibleUsername". Kicks, bans, and warnings all work as expected with no errors whatsoever.

Another note on 2): config.yml
```
  # If enabled, all active warnings will be deactivated after a warning action is executed.
  delete_warnings_after_action: false
```

**Comment by mibby** [comment]
@ruany Here is an archived test environment so you can reproduce the issue. Please remove the link as soon as you have obtained it.
https://www.dropbox.com/s/c74m0hskz5j5z0c/Server.zip?dl=0

* Launch start.bat in BungeeCord folder.
* Launch start.bat in Server folder.
* Connect to server via "localhost" or ":25565"
* Warn yourself from server console via
```warn <name> test1```
```warn <name> test2```
```warn <name> test3```
* You aren't kicked from the server or BungeeCord after the 3 warnings,, nor are banned after receiving the 3 warnings when warned from console.

Names that are 13 characters or more produce the error reported above. Removing all Essentials plugin jars from the server, warnings properly kick and ban. So it must be a conflict between Essentials and LiteBans when one of those Essentials settings I posted above are enabled.

**Comment by ruany** [comment]
Ok, I've executed your test environment and can replicate the issue.

The bug was in LiteBans and has been fixed in version 2.2.7 which I've just uploaded.

* Fixed warning actions being executed on display names instead of regular names when `use_display_names` is enabled in config.yml.

(My test environment didn't have `use_display_names` enabled in LiteBans' config.yml. I vaguely remember enabling it at some point, but I've reset my configuration since and forgot about it)
